spec/ddtrace/integration_spec.rb

```diff
@@ -109,33 +109,75 @@ def agent_receives_span_step3(previous_success)
     end
   end
 
-  describe 'shutdown executes only once' do
+  describe 'shutdown' do
     include_context 'agent-based test'
 
-    before(:each) do
-      tracer.trace('my.short.op') do |span|
-        span.service = 'my.service'
-      end
+    context 'executes only once' do
+      subject(:multiple_shutdown) do
+        tracer.trace('my.short.op') do |span|
+          span.service = 'my.service'
+        end
 
-      mutex = Mutex.new
-      @shutdown_results = []
+        threads = Array.new(10) do
+          Thread.new { tracer.shutdown! }
+        end
 
-      threads = Array.new(10) do
-        Thread.new { mutex.synchronize { @shutdown_results << tracer.shutdown! } }
+        threads.each(&:join)
       end
 
-      threads.each(&:join)
+      let(:stats) { tracer.writer.stats }
+
+      it do
+        multiple_shutdown
+        expect(stats[:traces_flushed]).to eq(1)
+        expect(stats[:services_flushed]).to be_nil
+        expect(stats[:transport].client_error).to eq(0)
+        expect(stats[:transport].server_error).to eq(0)
+        expect(stats[:transport].internal_error).to eq(0)
+      end
     end
 
-    let(:stats) { tracer.writer.stats }
+    context 'when sent TERM' do
+      subject(:terminated_process) do
+        # Initiate IO pipe
+        pipe
 
-    it do
-      expect(@shutdown_results.count(true)).to eq(1)
-      expect(stats[:traces_flushed]).to eq(1)
-      expect(stats[:services_flushed]).to be_nil
-      expect(stats[:transport].client_error).to eq(0)
-      expect(stats[:transport].server_error).to eq(0)
-      expect(stats[:transport].internal_error).to eq(0)
+        # Fork the process
+        fork_id = fork do
+          allow(Datadog.tracer).to receive(:shutdown!).and_wrap_original do |m, *args|
+            m.call(*args).tap { write.write(graceful_signal) }
+          end
+
+          tracer.trace('my.short.op') do |span|
+            span.service = 'my.service'
+          end
+
+          sleep(1)
+        end
+
+        # Give the fork a chance to setup and sleep
+        sleep(0.2)
+
+        # Kill the process
+        write.close
+        Process.kill('TERM', fork_id) rescue nil
```

Comment:
When what scenarios would you expect `Process.kill` to raise an error here?

Reply:
Maybe if the process already completed, not entirely sure. Mostly was trying to recreate the conditions from https://github.com/resque/resque/blob/master/lib/resque/worker.rb#L462

Reply:
The process finishing before we can send the signal seems plausible.

Reply:
In this test that shouldn't happen, given how long it sleeps. Either way, I think it should be safe.